File: generator/regions.py
import boto3
import util
from typing import Dict, Set

REGIONS: Set[str] = set()
SESSIONS: Dict[str, boto3.Session] = {}

account = boto3.client("account")

def GetRegions(*, next_token=""):
    if len(REGIONS) > 0 and not next_token:
        return REGIONS
    response = account.list_regions(MaxResults=50) if next_token else account.list_regions()
    for region in response["Regions"]:
        REGIONS.add(region["RegionName"])
    if "NextToken" in response:
        GetRegions(next_token=response["NextToken"])
        return
    
    util.WriteFile("../data/regions.json", REGIONS)
    return REGIONS

def GetSessions() -> Dict[str, boto3.Session]:
    GetRegions()
    for region in list(REGIONS):
        SESSIONS[region] = boto3.Session(region_name=region)
    return SESSIONS

File: generator/lib/regions.ts
import { AccountClient, paginateListRegions, EnableRegionCommand, ListRegionsCommand, RegionOptStatus } from '@aws-sdk/client-account';
import { writeDataToFile } from './util';

// Create a function that will get all regions for an account
export async function getAllRegions(): Promise<string[]> {
  const accountClient = new AccountClient({ region: 'us-east-1' });
  const regions: string[] = [];
  for await (const page of paginateListRegions({ client: accountClient }, {})) {
    regions.push(
      ...page.Regions!.map((r) => {
        if (r.RegionOptStatus == "DISABLED") {
          const enableCommand = new EnableRegionCommand({ RegionName: r.RegionName! });
          accountClient.send(enableCommand);
        }
        return r.RegionName!;
      })
    );
  }
  const listCommand = new ListRegionsCommand({RegionOptStatusContains: [RegionOptStatus.ENABLING]});
  for (let output = await accountClient.send(listCommand); output.Regions!.length > 0; output = await accountClient.send(listCommand)) {
    console.log("Regions still being enabled: ", output.Regions);
    await new Promise(r => setTimeout(r, 5000));
  }
  writeDataToFile(
    "src/lib/regions.json",
    regions,
  );
  return regions;
}
